docs/source/configuration.rst
Configuration
-------------

AsyncHBase 1.7 has a new configuration class ``Config`` that enables tweaking many parameters of the client. For backwards compatibility, the existing ``HBaseClient`` constructors will instantiate a new config object with default values. However new constructors will accept a config object.

To create a config with default values, instantiate it via ``Config config = new Config();``. Or to load properties from a file, call ``Config config = new Config(<path_to_file>);`` where the file is a standard Java properties file with entries like ``hbase.rpc.timeout = 60000`` separated by new lines.

To modify the (default or parsed from file) value of a configuration entry, simply call ``config.overrideConfig(<property>, <value>);``. Make sure to modify values *before* passing the config to the ``HBaseClient`` constructor. Modifying values of the config after instantiating the client will not cause an effect as many values are read from the config and stored in the client's memory, never to be read again.

Properties
^^^^^^^^^^

The following is a table of configuration options for AsyncHBase. Some of the values share names in common with HBase's configs but may behave slightly differently. Be sure to read the notes associated with the config before proceeding.

.. csv-table::
   :header: "Property", "Type", "Description", "Default"
   :widths: 20, 5, 65, 10

   "hbase.client.retries.number", "Integer", "The number of times an RPC will be retried when a recoverable exception occurs, e.g. a region moving or splitting. Non-recoverable exceptions will fail the RPC immediately.", "10"
   "hbase.increments.buffer_size", "Integer", "How many unique counters to maintain in memory when using ``BufferedIncrement`` RPCs. Once this limit is reached, new buffered increments are immediately sent to the region server.", "65,535"
   "hbase.increments.durable", "Boolean", "Whether or not to write to the WAL for every atomic increment request.", "true"
   "hbase.ipc.client.connection.idle_timeout", "Integer", "How long, in seconds, to wait before closing an idle connection to a region server. Idle means no writes or reads to the socket.", "300"
   "hbase.ipc.client.socket.receiveBufferSize", "Integer", "The size of the region server socket receive buffer in bytes.", "System dependent"
   "hbase.ipc.client.socket.sendBufferSize", "Integer", "The size of the region server socket send buffer in bytes.", "System dependent"
   "hbase.ipc.client.socket.timeout.connect", "Integer", "How long, in milliseconds, to wait for a socket connection attempt to a region server.", "5000"
   "hbase.ipc.client.tcpkeepalive", "Boolean", "Whether or not to enable keep-alives on the TCP socket to the region servers.", "true"
   "hbase.ipc.client.tcpnodelay", "Boolean", "Whether or not to bypass TCP spooling and send packets immediately to the region servers.", "true"
   "hbase.kerberos.regionserver.principal", "String", "A name or template to use as the principal for kerberos authentication. Instances of the string ``_HOST`` will be replaced with the host name of the client.", ""
   "hbase.meta.scan", "Boolean", "Whether or not to force scanning for region server locations in Meta instead of using the older getClosestRowBefore() method to lookup a region. This is required for pre-release versions of HBase 2.0. For later versions of 2.x, HBase will throw an exception caught by AsyncHBase 1.8.2 and the client will automatically switch to scanning.", "false"
   "hbase.meta.split", "Boolean", "Whether or not experimental split-meta has been enabled on the HBase cluster.", "false"
   "hbase.nsre.high_watermark", "Integer", "The maximum number of RPCs allowed in the NSRE queue for a region before new RPCs are failed with please throttle exceptions.", "10000"
   "hbase.nsre.low_watermark", "Integer", "The number of RPCs, per region server, that must be present in the NSRE queue before warnings start appearing in the log file.", "1000"
   "hbase.region_client.check_channel_write_status", "Boolean", "Whether or not to check the write status of the region server connection's socket before attempting to send the RPC. If a region server is slow, the client may buffer a large amount of data in memory while waiting for the server to consume data. This can lead to out of memory issues or a very large heap. When this is enabled and the buffer is full, RPCs will fail with please throttle exceptions.", "false"
   "hbase.region_client.inflight_limit", "Integer", "The maximum number of RPCs sent to the connected region server and waiting a response. A large number of inflight RPCs can lead to out of memory issues. A value of 0 will disable this limit. If the limit is reached, RPCs will fail with please throttle exceptions.", "0"
   "hbase.region_client.pending_limit", "Integer", "The maximum number of RPCs queued while waiting for a region server connection to complete. If a region server is slow to respond to a connection request, or authentication takes a long time, then a large queue of RPCs may build up in memory. A value of 0 will disable this limit. If the limit is reached, RPCs will fail with please throttle exceptions.", "0"
   "hbase.regionserver.kerberos.password", "String", "NOT RECOMMENDED OR TESTED: A password for the region server when authenticating via kerberos.", ""
   "hbase.rpcs.batch.size", "Integer", "The number of individual RPCs that can be buffered before being sent to the region server. If this limit is reached before ``hbase.rpcs.buffered_flush_interval`` then the batch is sent out immediately and a new batch is started.", "1024"
   "hbase.rpcs.buffered_flush_interval", "Integer", "How often, in milliseconds, to flush buffered RPCs (Puts, Appends, etc) to the region servers.", "1000"
   "hbase.rpc.protection", "String", "Whether or not to encrypt RPCs while they are traversing the network. May be ``authentication`` (no encryption), ``integrity``, (no encryption) or ``privacy`` (encrypted). Requires authentication to be enabled."
   "hbase.rpc.timeout", "Integer", "How long, in milliseconds, to wait for a response to an RPC from a region server before failing the RPC with a ``RpcTimedOutException``. This value can be overridden on a per-RPC basis. A value of 0 will not allow RPCs to timeout.", "0"
   "hbase.sasl.clientconfig", "String", "The section in a JAAS configuration file to use when authenticating against a region server.", "Client"
   "hbase.security.auth.94", "Boolean", "Whether or not the client is connection to an HBase 0.94 cluster with authentication.", "false"
   "hbase.security.auth.enable", "Boolean", "Whether or not to enable authentication when connecting to HBase. Please see :doc:`authentication` for more information."
   "hbase.security.authentication", "String", "The type of authentication required for the HBase cluster. May be ``kerberos`` or ``simple``.", ""
   "hbase.security.simple.username", "String", "The user name for authenticating against an HBase cluster with simple authentication. (Not recommended)", ""
   "hbase.timer.tick", "Integer", "How often, in milliseconds, to execute the flush and timeout timers, looking for new tasks. This value should not need modification.", "20"
   "hbase.timer.ticks_per_wheel", "Integer", "How many hash buckets are used for scheduling tasks in the flush and timeout timers. This value should not need modification.", "512"
   "hbase.workers.size", "Integer", "The number of worker threads to instantiate for region client connections.", "2 * CPU cores"
   "hbase.zookeeper.getroot.retry_delay", "Integer", "How long, in milliseconds, to wait between attempts fetching the root region from Zookeeper if a call fails.", "1000"
   "hbase.zookeeper.quorum", "String", "A comma-separated list of ZooKeeper hosts to connect to, with or without port specifiers.
   E.g. ``192.168.1.1:2181,192.168.1.2:2181``", "localhost"
   "hbase.zookeeper.session.timeout", "Integer", "How long, in milliseconds, to maintain a Zookeeper connection. After the session times out, the connection is closed and a new one will be opened the next time AsyncHBase needs root.", "5000"
   "hbase.zookeeper.znode.parent", "String", "Path under which the znode for the -ROOT- region is located", "/hbase"

.. NOTE::

  For authentication with Kerberos, a JAAS file must be created and passed to the JVM using AsyncHBase in the system parameter ``java.security.auth.login.config``.

Data Types
^^^^^^^^^^

Some configuration values require special consideration:

* Booleans - The following literals will parse to ``True``:

  * ``1``
  * ``true``
  * ``yes``
  
  Any other values will result in a ``False``. Parsing is case insensitive
  
* Strings - Strings, even those with spaces, do not require quotation marks, but some considerations apply:

  * Special characters must be escaped with a backslash include: ``#``, ``!``, ``=``, and ``:``
    E.g.::
    
      my.property = Hello World\!
      
  * Unicode characters must be escaped with their hexadecimal representation, e.g.::
  
      my.property = \u0009


NEWS
AsyncHBase - User visible and noteworthy changes.
This project uses Semantic Versioning (see http://semver.org/).

* Version 1.8.2 (2018-05-19)

This is a bug fix release with one beta feature.
  - Fix a possible issue with the security config when spaces are present in the
    authentication type.
  - Add beta support for split-meta from HBASE-11288 for large, multi-tenant
    clusters. The table names may change in the future.
  - Auto-detect HBase 2.0 by catching an exception thrown when getClosestRowOrBefore
    is called. HBASE-20237.
  - Log the version setting, min and max timestamps in Scanners.
  - Add CallQueueTooBig retry handling in the RegionClient without NSREing.

* Version 1.8.1 (2018-03-18)

This is a bug fix release.
  - Revert locking in an RPC to synchronizing on the key instead of the entire
    object as well as setting the suspend probe flag to volatile.
  - Fix GetRequests where the family was masked and immutable after being set.
  - Fix setting the scanner reverse flag in the setReversed() setter. 
    Previously when called it would only set the flag to `true`.
  - Add RegionTooBusyException to allow for retries when newer versions of HBase
    send this exception.
  - Fix processing of encoded (private) RPCs and catch + log exceptions during
    SASL negotiation.
  - Add a configuration flag 'hbase.meta.scan' to switch meta lookups to 
    use a reverse scan when finding tables for compatibility with HBase 2.x
	where the old method of getting the closest row was disabled.
  - Fix an issue with AssertionErrors being thrown when a multi-action RPC
    had an exception of a type that didn't implement the make() method.
  - Fix an issue with multi-actions sent to HBase 1.31 an later where the
    request was sorted by row key so that responses no longer matched the
	request order. Now we sort by row key as well.
  
Noteworthy bug Fixes:

* Version 1.8.0 (2017-06-09)

This is feature release with new APIs and behaviors.

New Features:
  - Support HBase 1.3.0 and later where scanner behavior changed.
  - Add optional support for buffered AtomicIncrement calls to bypass the WAL
    for quicker writes.
  - Add an API for writing AtomicIncrements over multiple columns in a single
    RPC.
  - Add the MultipleColumnPrefixFilter.
  - Add a "multi-get" API that allows for batching GetRequests and sending them
    in a single request. Useful if the client wants to fetch a number of 
    disparate rows when scanning would be less efficient.
  - Support for reverse scanning

* Version 1.7.3 (2017-06-09)

This is a bug fix release.

Noteworthy bug Fixes:
  - Properly handle IPV6 addresses in the RegionClientPipeline.
  - Avoid a deadlock when calling the RegionClient.toString() method.
  - Avoid a synchronization deadlock handling NSRE probes.

* Version 1.7.2 (2016-09-17)

This is a bug fix release.

Noteworthy bug Fixes:
  - Fix a race condition where messages in the NSRE queue could be stuck and
    never retried properly. This will help region server crash recovery.
  - Fix a bug in batched RPCs where, on exception, the batch RPC itself was
    returned in the exception instead of the individual RPCs.
  - Changed watermark variables to integers instead of shorts to avoid 
    overflows when set to values over 32K.
  - Added retries of CallQueueTooBigExceptions. These still need to avoid
    flushing the region cache but at least they'll be retried.
  - Fix calls to scanner.setFamilies() where NPEs were being thrown.
  - Retry RPCs on RegionServerAbortedException, ServerNotRunningYetExceptions
    and RegionServerStoppedExceptions.
  - Fix CompareAndSetRequests so they serialize timestamps as well as the data
    to target specific versions.
  - Update the protobuf compiler plugin for Maven.

* Version 1.7.1 (2016-02-10)

This is a bug fix release.

Noteworthy bug Fixes:
  - Fix a regression where RPCs from pre-0.96 HBase clusters were not parsed
    properly when large responses were split across packets. RPC timeouts were
    not handled properly either and now they are.

* Version 1.7.0 (2015-09-11) 

This release adds support for secure HBase installations, more filters and
greater configuration.

New public APIs:
  - The Config object handles configuration of the client and can be set
    manually or it may load a list of settings from a properties file.
  - Introduction of AppendRequest to allow for appending data to the value
    of a column in HBase. Appends can be buffered similar to PutRequests.
  - Introduction of Kerberos and simple authentication support against HBase
    region servers. Configuration is similar to the HTable client.
  - Introduction of locateRegions() and RegionLocation similar to HTable to
    allow users the ability to get a list of regions for a table.
  - Add filter support and timestamp ranges to GetRequests.
  - Add a socket idle handler to close connections to region clients after
    some period of inactivity.
  - Introduce RPC timeouts on either a global or per-RPC basis.
  - Netty threads are now named for easier debugging.
  - Add filters including FirstKeyOnlyFilter, ColumnPaginationFilter,
    KeyOnlyFilter and FuzzyRowFilter
  - Introduce the RegionClientStats object for tracking metrics around individual
    region client connections.

Noteworthy bug fixes:
  - Add the RegionOpeningException as a recoverable exception (#117)
  - Region clients can now check the status of the socket write buffer before
    sending RPCs to avoid OOM issues. Must be enabled via the config.
  - Fix a small meta lookup permit bug when an exception is thrown and the 
    semaphore isn't released.
  - Fix an issue with stale region clients created when a region server
    restarts.
  - Fix a bug with CDH5 and HBase 0.96 when closing a scanner that has rows
    remaining to be read.
  - Fix a potential bug with decoding byte array backed ChannelBuffers in
    the region client.

src/Config.java
/*
 * Copyright (C) 2015  The Async HBase Authors.  All rights reserved.
 * This file is part of Async HBase.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *   - Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *   - Redistributions in binary form must reproduce the above copyright notice,
 *     this list of conditions and the following disclaimer in the documentation
 *     and/or other materials provided with the distribution.
 *   - Neither the name of the StumbleUpon nor the names of its contributors
 *     may be used to endorse or promote products derived from this software
 *     without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package org.hbase.async;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * A configuration class that stores key/value pairs as strings similar to
 * the System properties object. The class includes helper methods to parse the
 * strings into numbers or booleans and it allows for loading a standard Java
 * properties file from disk. The  {@link loadSystemAndDefaults} will be called
 * to set defaults and optionally load values from the System properties list.
 * <p>
 * Note that multiple threads can read from this config object but modifying
 * values is not thread safe.
 * @since 1.7.0
 */
public class Config {
  private static final Logger LOG = LoggerFactory.getLogger(Config.class);

  /** Flag to determine if we're running under Windows or not */
  public static final boolean RUNNING_WINDOWS;
  static {
    // this should never be null, but who knows?
    RUNNING_WINDOWS = System.getProperty("os.name") != null ? System
        .getProperty("os.name").contains("Windows") : false;
  }
  
  /**
   * The list of properties configured to their defaults or modified by users.
   * We use a non-synchronized hash map for fast access as opposed to the
   * HashTable backed properties objects.
   */
  protected final HashMap<String, String> properties = new HashMap<String, String>();

  /** Holds default values for the config */
  protected final HashMap<String, String> default_map = new HashMap<String, String>();

  /** Tracks the location of the file that was actually loaded */
  protected String config_location;

  /**
   * Constructor that initializes default configuration values from the
   * System.properties map.
   */
  public Config() {
    loadSystemAndDefaults();
  }

  /**
   * Constructor that initializes default values and attempts to load the given
   * properties file
   * @param file Path to the file to load
   * @throws IOException Thrown if unable to read or parse the file
   */
  public Config(final String file) throws IOException {
    loadSystemAndDefaults();
    loadConfig(file);
  }

  /**
   * Constructor for subclasses who want a copy of the parent config but without
   * the ability to modify the original values.
   * 
   * This constructor will not re-read the file, but it will copy the location
   * so if a child wants to reload the properties periodically, they may do so
   * @param parent Parent configuration object to load from
   */
  public Config(final Config parent) {
    properties.putAll(parent.properties);
    config_location = parent.config_location;
    loadSystemAndDefaults();
  }

  /**
   * Allows for modifying properties after loading
   * 
   * WARNING: This should only be used on initialization of the config object
   * and it is not thread safe.
   * 
   * @param property The name of the property to override
   * @param value The value to store
   */
  public void overrideConfig(final String property, final String value) {
    properties.put(property, value);
  }

  /**
   * Returns the given property as a String
   * @param property The property to fetch
   * @return The property value as a string or null if the property did not exist
   */
  public final String getString(final String property) {
    return properties.get(property);
  }

  /**
   * Returns the given property as an integer
   * @param property The property to fetch
   * @return A parsed integer or an exception if the value could not be parsed
   * @throws NumberFormatException if the property could not be parsed
   * @throws NullPointerException if the property did not exist
   */
  public final int getInt(final String property) {
    return Integer.parseInt(properties.get(property));
  }

  /**
   * Returns the given property as a short
   * @param property The property to fetch
   * @return A parsed short or an exception if the value could not be parsed
   * @throws NumberFormatException if the property could not be parsed
   * @throws NullPointerException if the property did not exist
   */
  public final short getShort(final String property) {
    return Short.parseShort(properties.get(property));
  }

  /**
   * Returns the given property as a long
   * @param property The property to fetch
   * @return A parsed long or an exception if the value could not be parsed
   * @throws NumberFormatException if the property could not be parsed
   * @throws NullPointerException if the property did not exist
   */
  public final long getLong(final String property) {
    return Long.parseLong(properties.get(property));
  }

  /**
   * Returns the given property as a float
   * @param property The property to fetch
   * @return A parsed float or an exception if the value could not be parsed
   * @throws NumberFormatException if the property could not be parsed
   * @throws NullPointerException if the property did not exist
   */
  public final float getFloat(final String property) {
    return Float.parseFloat(properties.get(property));
  }

  /**
   * Returns the given property as a double
   * @param property The property to fetch
   * @return A parsed double or an exception if the value could not be parsed
   * @throws NumberFormatException if the property could not be parsed
   * @throws NullPointerException if the property did not exist
   */
  public final double getDouble(final String property) {
    return Double.parseDouble(properties.get(property));
  }

  /**
   * Returns the given property as a boolean
   * Property values are case insensitive and the following values will result
   * in a True return value: - 1 - True - Yes
   * Any other values, including an empty string, will result in a False
   * @param property The property to fetch
   * @return A parsed boolean
   * @throws NullPointerException if the property was not found
   */
  public final boolean getBoolean(final String property) {

    if (properties.containsKey(property)) {
      final String val = properties.get(property).toUpperCase();

      if (val.equals("1"))
        return true;
      if (val.equals("TRUE"))
        return true;
      if (val.equals("YES"))
        return true;
    }

    return false;
  }

  /**
   * Returns the directory name, making sure the end is an OS dependent slash
   * @param property The property to fetch
   * @return The property value with a forward or back slash appended
   * @throws NullPointerException if the property was not found
   */
  public final String getDirectoryName(final String property) {
    final String directory = properties.get(property);
    if (directory == null || directory.isEmpty()) {
      return null;
    }
    if (RUNNING_WINDOWS) {
      // Windows swings both ways. If a forward slash was already used,
      // we'll
      // add one at the end if missing. Otherwise use the windows default
      // of \
      if (directory.charAt(directory.length() - 1) == '\\'
          || directory.charAt(directory.length() - 1) == '/') {
        return directory;
      }
      if (directory.contains("/")) {
        return directory + "/";
      }
      return directory + "\\";
    }
    
    if (directory.contains("\\")) {
      throw new IllegalArgumentException(
          "Unix path names cannot contain a back slash");
    }
    if (directory.charAt(directory.length() - 1) == '/') {
      return directory;
    }
    return directory + "/";
  }

  /**
   * Determines if the given property is in the map
   * @param property The property to search for
   * @return True if the property exists and has a value, not an empty string
   */
  public final boolean hasProperty(final String property) {
    final String val = properties.get(property);
    if (val == null)
      return false;
    if (val.isEmpty())
      return false;
    return true;
  }

  /**
   * Returns a simple string with the configured properties for debugging. Note
   * that any property with the string "PASS" will be obfuscated to hide
   * passwords.
   * @return A string with information about the config
   */
  public final String dumpConfiguration() {
    if (properties.isEmpty()) {
      return "No configuration settings stored";
    }

    final StringBuilder response = new StringBuilder("Configuration:\n");
    response.append("File [" + this.config_location + "]\n");
    int line = 0;
    for (Map.Entry<String, String> entry : properties.entrySet()) {
      if (line > 0) {
        response.append("\n");
      }
      response.append("Key [" + entry.getKey() + "]  Value [");
      if (entry.getKey().toUpperCase().contains("PASS")) {
        response.append("********");
      } else {
        response.append(entry.getValue());
      }
      response.append("]");
      line++;
    }
    return response.toString();
  }

  /** @return An copy of the configuration map */
  public final Map<String, String> getMap() {
    return new HashMap<String, String>(properties);
  }

  /**
   * Loads options from the JVM system properties and/or sets defaults
   */
  private void loadSystemAndDefaults() {
    /* --------- AsyncHBase specific configs --------- */
    default_map.put("hbase.rpcs.buffered_flush_interval", "1000");
    default_map.put("hbase.rpcs.batch.size", "1024");
    default_map.put("hbase.region_client.check_channel_write_status", "false");
    default_map.put("hbase.region_client.inflight_limit", "0");
    default_map.put("hbase.region_client.pending_limit", "0");
    default_map.put("hbase.nsre.event_queue_size", "128");
    default_map.put("hbase.nsre.low_watermark", "1000");
    default_map.put("hbase.nsre.high_watermark", "10000");
    default_map.put("hbase.timer.tick", "20");
    default_map.put("hbase.timer.ticks_per_wheel", "512");
    default_map.put("hbase.security.auth.enable", "false");
    default_map.put("hbase.zookeeper.getroot.retry_delay", "1000");
    default_map.put("hbase.hbase.ipc.client.connection.idle_timeout", "300");
    
    /**
     * How many different counters do we want to keep in memory for buffering.
     * Each entry requires storing the table name, row key, family name and
     * column qualifier, plus 4 small objects.
     *
     * Assuming an average table name of 10 bytes, average key of 20 bytes,
     * average family name of 10 bytes and average qualifier of 8 bytes, this
     * would require 65535 * (10 + 20 + 10 + 8 + 4 * 32) / 1024 / 1024 = 11MB
     * of RAM, which isn't too excessive for a default value.  Of course this
     * might bite people with large keys or qualifiers, but then it's normal
     * to expect they'd tune this value to cater to their unusual requirements.
     */
    default_map.put("hbase.increments.buffer_size", "65535");
    
    /* --- HBase configs (same names as their HTable counter parts --- 
     * Note that the defaults may differ though */
    default_map.put("hbase.zookeeper.quorum", "localhost");
    // HBase drops the "hbase" bit. *shrug*
    default_map.put("hbase.zookeeper.znode.parent", "/hbase");
    default_map.put("hbase.zookeeper.session.timeout", "5000");
    default_map.put("hbase.client.retries.number", "10");
    /** Note that HBase's client defaults to 60 seconds. We default to 0 for
     * AsyncHBase backwards compatibility. This may change in the future.
     */
    default_map.put("hbase.rpc.timeout", "0");
    default_map.put("hbase.ipc.client.connection.maxidletime", "0");
    default_map.put("hbase.ipc.client.connect.max.retries", "0");
    default_map.put("hbase.ipc.client.socket.timeout.connect", "5000");
    default_map.put("hbase.ipc.client.tcpnodelay", "true");
    default_map.put("hbase.ipc.client.tcpkeepalive", "true");
    
    // default rate limiter values
    // Maximum write rate restriction per second, there will not be any 
    // restriction after that
    default_map.put("hbase.rpc.ratelimit.rate_max_threshold", "1000");//per sec
    // Minimum allowed write rate restriction per second
    default_map.put("hbase.rpc.ratelimit.rate_min_threshold", "10");//per sec
    // Write rate will be decreased or increased by this value, in percentage
    default_map.put("hbase.rpc.ratelimit.rate_of_change", "10");//in %
    // Region server health will be analyzed again and recalculate write rate
    // after this time frame, in milli seconds
    default_map.put("hbase.rpc.ratelimit.time_frame", "60000");//in ms
    // Enable or disable write limit
    default_map.put("hbase.rpc.ratelimit.enable", "false");

    // Minimum number of responses required to say the region is healthy, it is
    // proportional to the number of write attempts within that time frame
    default_map.put("hbase.rpc.ratelimit.policy.min_success_rate", "50");//in %
    // Minimum number of write attempts required for rate limit calculation
    default_map.put("hbase.rpc.ratelimit.policy.min_writes_req", "1000");
    
    for (Map.Entry<String, String> entry : default_map.entrySet()) {
      if (!properties.containsKey(entry.getKey()))
        properties.put(entry.getKey(), entry.getValue());
    }
  }
  
  /**
   * Attempts to load the configuration from the given location
   * @param file Path to the file to load
   * @throws IOException Thrown if there was an issue reading the file
   * @throws FileNotFoundException Thrown if the config file was not found
   */
  protected void loadConfig(final String file) throws FileNotFoundException,
      IOException {
    final FileInputStream file_stream = new FileInputStream(file);
    try {
      final Properties props = new Properties();
      props.load(file_stream);
  
      // load the hash map
      loadHashMap(props);
  
      // no exceptions thrown, so save the valid path and exit
      LOG.info("Successfully loaded configuration file: " + file);
      config_location = file;
    } finally {
      file_stream.close();
    }
  }

  /**
   * Called from {@link #loadConfig} to copy the properties into the hash map.
   * @param props The loaded Properties object to copy
   */
  private void loadHashMap(final Properties props) {
    @SuppressWarnings("rawtypes")
    final Enumeration e = props.propertyNames();
    while (e.hasMoreElements()) {
      final String key = (String) e.nextElement();
      properties.put(key, props.getProperty(key));
    }
  }

}